master/internal/resourcemanagers/priority_test.go

```diff
@@ -0,0 +1,271 @@
+
+	agents := []*mockAgent{
+		{id: "agent1", slots: 4},
+		{id: "agent2", slots: 4},
+	}
+	groups := []*mockGroup{
+		{id: "group1", priority: &lowerPriority},
+		{id: "group2", priority: &higherPriority},
+	}
+	tasks := []*mockTask{
+		{id: "task1", slotsNeeded: 4, group: groups[0]},
+		{id: "task2", slotsNeeded: 1, group: groups[0]},
+		{id: "task3", slotsNeeded: 12, group: groups[1]},
+	}
+
+	system := actor.NewSystem(t.Name())
+	taskList, groupMap, agentMap := setupSchedulerStates(t, system, tasks, groups, agents)
+
+	p := &priorityScheduler{}
+	toAllocate, _ := p.prioritySchedule(taskList, groupMap, agentMap, BestFit)
+
+	expectedToAllocate := []*mockTask{}
+	assertEqualToAllocate(t, toAllocate, expectedToAllocate)
+
+	// Check that agent stat has not changed.
+	for _, agent := range agentMap {
+		assert.Equal(t, agent.numEmptySlots(), 4)
+	}
+}
+
+func TestPrioritySchedulingNoPreemptionWithLabels(t *testing.T) {
+	lowerPriority := 50
+	higherPriority := 40
+
+	agents := []*mockAgent{
+		{id: "agent1", slots: 4, label: "label1"},
+		{id: "agent2", slots: 4, label: "label1"},
+	}
+	groups := []*mockGroup{
+		{id: "group1", priority: &lowerPriority},
+		{id: "group2", priority: &higherPriority},
+	}
+	tasks := []*mockTask{
+		{id: "task1", slotsNeeded: 4, group: groups[0], label: "label1"},
+		{id: "task2", slotsNeeded: 1, group: groups[0], label: "label1"},
+		{id: "task3", slotsNeeded: 4, group: groups[1], label: "label2"},
+	}
+
+	system := actor.NewSystem(t.Name())
+	taskList, groupMap, agentMap := setupSchedulerStates(t, system, tasks, groups, agents)
+
+	p := &priorityScheduler{}
+	toAllocate, _ := p.prioritySchedule(taskList, groupMap, agentMap, BestFit)
+
+	expectedToAllocate := []*mockTask{tasks[0], tasks[1]}
+	assertEqualToAllocate(t, toAllocate, expectedToAllocate)
+}
+
+func TestPrioritySchedulingPreemption(t *testing.T) {
+	lowerPriority := 50
+	higherPriority := 40
+
+	agents := []*mockAgent{
+		{id: "agent1", slots: 4},
+	}
+	groups := []*mockGroup{
+		{id: "group1", priority: &lowerPriority},
+		{id: "group2", priority: &higherPriority},
+	}
+	tasks := []*mockTask{
+		{id: "task1", slotsNeeded: 4, group: groups[0],
+			allocatedAgent: agents[0], containerStarted: true},
+		{id: "task2", slotsNeeded: 0, group: groups[0]},
+		{id: "task3", slotsNeeded: 4, group: groups[1]},
+	}
+
+	system := actor.NewSystem(t.Name())
+	taskList, groupMap, agentMap := setupSchedulerStates(t, system, tasks, groups, agents)
+
+	for _, agent := range agentMap {
+		fmt.Println("agent has free slots: ", agent.numEmptySlots())
+	}
+
+	p := &priorityScheduler{preemptionEnabled: true}
+	toAllocate, toRelease := p.prioritySchedule(taskList, groupMap, agentMap, BestFit)
+
+	expectedToAllocate := []*mockTask{tasks[1]}
+	assertEqualToAllocate(t, toAllocate, expectedToAllocate)
+
+	expectedToRelease := []*mockTask{tasks[0]}
+	assertEqualToRelease(t, taskList, toRelease, expectedToRelease)
+}
+
+func TestPrioritySchedulingNoSchedule(t *testing.T) {
```

Comment:
q: why is this the expected behavior? 
blocking: the test function name should be more descriptive

Reply:
Because the higher priority task can not be scheduled, but we test to make sure the lower priority task which can be scheduled is blocked due to priorities. 

What name would you suggest?

Reply:
Haha this is a pretty long function name but I like it.